File: MinimumRotatedArray.java
//Time : O(log n) n is number of array
//Space : O(1)
//Solved on Leet code :Yes
// Approach : binary search, move towards unsorted section

class MinimumRotatedArray {
    public int findMin(int[] nums) {
        int low = 0;
        int high = nums.length-1;
        
        while(low<=high){
            
            if(nums[low] <= nums[high]){
                return nums[low];
            }
            
            int mid = low + (high - low)/2;
            
            if ( (mid==0 || nums[mid - 1] > nums[mid]) && ( mid == nums.length-1 ||  nums[mid + 1] > nums[mid])){
                return nums[mid];
                }
            
            
            if(nums[low] <= nums[mid]){
                low = mid + 1;
            }else{
                high = mid -1;
            }
            
        }
        return 123123;
    }

    public static void main(String[] args){
        MinimumRotatedArray minimumRotatedArray = new MinimumRotatedArray();
        System.out.println(minimumRotatedArray.findMin(new int[] {2,4,1}));
    }
}

File: minimumrotatedarray.py
#Time Complexity: O(log n)
#Space Complexity: O(1)
#Did your code executed on leetcode : yes

class Solution:
    def findMin(self, nums: List[int]) -> int:
        low = 0
        high = len(nums) - 1
        
        while (low <= high):
            
            #check whether entire array is sorted
            if nums[low] <= nums[high]: #<= because array might contain only one element
                return nums[low]
            
            #array is not fully sorted
            mid = low + (high - low)//2
            
            #check whether mid is smaller that previous and next element in list and here we are also going to check edge conditions
            
            if (mid == 0 or nums[mid - 1] > nums[mid]) and (mid == len(nums)-1 or nums[mid + 1] > nums[mid] ):
                return nums[mid]
            
            #if mid is not minimum then we check which part of array is sorted and minimum lies in unsorted part of array
            
            elif (nums[mid] >= nums[low]): #left side is sorted so move low
                low = mid + 1
            else:
                high = mid - 1
        
